File: android/AlpineApp/app/src/main/java/com/alpine/app/data/rpc/TlsConfig.kt
package com.alpine.app.data.rpc

import okhttp3.CertificatePinner
import okhttp3.OkHttpClient
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

enum class TlsMode {
    SYSTEM_CA,
    PINNED,
    TRUST_ALL
}

data class TlsConfig(
    val enabled: Boolean = false,
    val mode: TlsMode = TlsMode.SYSTEM_CA,
    val certFingerprint: String = "",
    val hostname: String = ""
) {
    fun applyTo(builder: OkHttpClient.Builder, host: String): OkHttpClient.Builder {
        if (!enabled) return builder

        return when (mode) {
            TlsMode.SYSTEM_CA -> builder

            TlsMode.PINNED -> {
                if (certFingerprint.isBlank()) return builder
                val pin = if (certFingerprint.startsWith("sha256/")) {
                    certFingerprint
                } else {
                    "sha256/$certFingerprint"
                }
                val pinner = CertificatePinner.Builder()
                    .add(hostname.ifBlank { host }, pin)
                    .build()
                builder.certificatePinner(pinner)
            }

            TlsMode.TRUST_ALL -> {
                val trustAllManager = object : X509TrustManager {
                    override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}
                    override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}
                    override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
                }
                val sslContext = SSLContext.getInstance("TLS").apply {
                    init(null, arrayOf<TrustManager>(trustAllManager), SecureRandom())
                }
                builder
                    .sslSocketFactory(sslContext.socketFactory, trustAllManager)
                    .hostnameVerifier { _, _ -> true }
            }
        }
    }

    fun buildUrl(host: String, port: String): String {
        val scheme = if (enabled) "https" else "http"
        return "$scheme://$host:$port"
    }
}

File: ios/AlpineApp/AlpineApp/Services/TlsConfig.swift
import Foundation

enum TlsMode: String, Codable, CaseIterable, Sendable {
    case systemCA = "SYSTEM_CA"
    case pinned = "PINNED"
    case trustAll = "TRUST_ALL"
}

struct TlsConfig: Sendable {
    var enabled: Bool = false
    var mode: TlsMode = .systemCA
    var certFingerprint: String = ""
    var backupFingerprints: [String] = []

    /// All fingerprints to check (primary + backups). Filters out empty strings.
    var allFingerprints: [String] {
        ([certFingerprint] + backupFingerprints).filter { !$0.isEmpty }
    }

    func buildURL(host: String, port: String) -> URL? {
        let scheme = enabled ? "https" : "http"
        return URL(string: "\(scheme)://\(host):\(port)/api/v1")
    }
}
